docs/guides/type_narrowing.rst

```diff
@@ -195,15 +195,91 @@ This behavior can be seen in the following example::
         else:
             reveal_type(x)  # Unrelated
 
+There are also cases beyond just mutability. In some cases, it may not be
```

Comment:
This fits better under "Safety and soundness" below, probably as a new subheader.

The example could be made more convincing; I think it's not going to be clear to most readers why it's unsafe.

I think something like this would demonstrate it:


```

def takes_any_a(a: A[int | str]):
    if possible_problem(a):
        assert_type(a, A[int])  # OK because A[int] is a subtype of A[int | str]
        if isinstance(a, B):
            assert_type(a, B[int])  # A[int] & B -> B[int]
            print(b.j + 1)  # boom
takes_any_a(B(i=1, j="x"))
```